### PR TITLE
Add TestWithDependenciesPlugin to build

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit adds a TestWithDependenciesPlugin to the gradle build. 
The main piece of functionality that it adds is to copy plugin-metadata 
from dependencies into the generated-resources for the current test source. 
This is necessary to ensure that permissions for dependencies are applied 
when running the tests.

A current limitation is that the permissions are applied differently 
than in the distribution sources. When permissions are granted to all 
depedencies for a module or plugin, the permissions are granted to all 
dependencies on the classpath for tests besides a few hardcoded 
exclusions:
- es core
- es test framework
- lucene test framework
- randomized runner
- junit library

### CODE REVIEW

**Comment by javanna** [comment]
can we have the name reflect what it does? I guess this thing does something other than being messy :)

**Comment by tbrooks8** [comment]
I named it following the pattern from the preexisting code. We already have a [non-rest](https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/groovy/org/elasticsearch/gradle/test/MessyTestPlugin.groovy#L35) version. So I just took that convention.

Is there something else you have in mind? RestTestWithDependenciesPlugin? RestTestWithESDependenciesPlugin? NonStandaloneRestTestPlugin?

**Comment by javanna** [comment]
> Is there something else you have in mind?

not really, everything you are proposing sounds good to me.

**Comment by rjernst** [comment]
Why do we need a new plugin? Can we just tweak the behavior of the existing standalone-rest-test plugin?

**Comment by tbrooks8** [comment]
Well - specifically I had modeled this PR on the relationship between MessyTestPlugin and StandaloneTestPlugin. It looks to me that MessyTestPlugin just extends StandaloneTestPlugin and adds the copying of dependency resources.

@nik9000 told me that there was more reasons for the original MessyTestPlugin.

One thing of note is that some modules which do not need the standalone-rest-test plugin (reindex) could use the functionality of copying resources of dependencies. Another approach that I had considered taking is not extending standalone-rest-test plugin and just creating a plugin to copy the resources. And that plugin can be applied as need be with no relation to other test plugins.

Or I could just modify the standalone-rest-test plugin. But then I would need a way to add this functionality eventually to tests that do not use that plugin.

**Comment by rjernst** [comment]
> And that plugin can be applied as need be with no relation to other test plugins.

I think this would be a good approach, as it would fit with the gradle model of making plugins to do one thing, and it would isolate the messiness.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/RestTestPlugin.groovy [hunk 1]
```diff
@@ -32,6 +32,7 @@ import org.gradle.api.Project
 public class RestTestPlugin implements Plugin<Project> {
     List REQUIRED_PLUGINS = [
         'elasticsearch.build',
+        'elasticsearch.rest-test-with-dependencies',
```
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/RestTestPlugin.groovy]: This should be `elasticsearch.test-with-dependencies`?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithDependenciesPlugin.groovy [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.elasticsearch.gradle.plugin.PluginBuildPlugin
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Dependency
+import org.gradle.api.artifacts.ProjectDependency
+import org.gradle.api.tasks.Copy
+
+/**
+ * A plugin to run messy REST tests. Messy tests are tests that depend on another
```
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithDependenciesPlugin.groovy]: Update javadocs? This is not longer just for rest tests?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/RestTestPlugin.groovy (hunk 1 above)
Comment (tbrooks8) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/RestTestPlugin.groovy]: Thanks for pointing this out. I actually think that "test-with-dependencies" should not be in that list anymore. As it has no relation to the RestTestPlugin.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithDependenciesPlugin.groovy (hunk 2 above)
Comment (tbrooks8) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/TestWithDependenciesPlugin.groovy]: I'll do that in just a minute. I'll also update the description of the PR to better reflect that changes.

**Review by rjernst — APPROVED** [review]

**Pull request merged**